client/asset/eth/eth.go

```diff
@@ -672,8 +672,41 @@ func (eth *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin
 
 // Redeem sends the redemption transaction, which may contain more than one
 // redemption.
-func (*ExchangeWallet) Redeem(form *asset.RedeemForm) ([]dex.Bytes, asset.Coin, uint64, error) {
-	return nil, nil, 0, asset.ErrNotImplemented
+func (eth *ExchangeWallet) Redeem(form *asset.RedeemForm) ([]dex.Bytes, asset.Coin, uint64, error) {
+	fail := func(err error) ([]dex.Bytes, asset.Coin, uint64, error) {
+		return nil, nil, 0, err
+	}
+
+	if len(form.Redemptions) == 0 {
+		return fail(errors.New("Redeem: must be called with at least 1 redemption"))
+	}
+
+	inputs := make([]dex.Bytes, 0, len(form.Redemptions))
+	var redeemedValue uint64
+	for _, redemption := range form.Redemptions {
+		var secretHash [32]byte
+		copy(secretHash[:], redemption.Spends.SecretHash)
+		if secretHash != sha256.Sum256(redemption.Secret) {
+			return fail(fmt.Errorf("Redeem: secretHash %x != sha256(%x)", secretHash, redemption.Secret))
+		}
+		swapData, err := eth.node.swap(eth.ctx, secretHash, form.AssetVersion)
+		if err != nil {
+			return nil, nil, 0, fmt.Errorf("Redeem: error finding swap state: %w", err)
+		}
+		redeemedValue += swapData.Value
+		inputs = append(inputs, redemption.Spends.Coin.ID())
+	}
+	outputCoin := eth.createAmountCoin(redeemedValue)
+	fundsRequired := dexeth.RedeemGas(len(form.Redemptions), form.AssetVersion) * form.FeeSuggestion
+
```

Comment:
This is where we should be calling `isRedeemable`, I believe.  The next step is to broadcast any number of secrets, so the txn better not fail.

Reply:
Yes this was the whole point haha, I've updated it